pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,271 @@
+			glog.Error(err)
+			errlist = append(errlist, err)
+		} else {
+			fullPathFile := path.Join(tempDir, fileName)
+			dir, _ := filepath.Split(fullPathFile)
+			if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+				return "", err
+			}
+			if e := ioutil.WriteFile(fullPathFile, []byte(sortStringfiedMap(data)), 0644); e != nil {
+				glog.Error(e)
+				errlist = append(errlist, err)
+			}
+		}
+	}
+	return tempDir, utilErrors.NewAggregate(errlist)
+}
+
+// updateSymLinks creates the relative links for the all the files
+func (m *metadataVolume) updateSymLinks() error {
+	for _, f := range m.fieldReferenceFileNames {
+		dir, _ := filepath.Split(f)
+		nbOfSubdir := 0
+		if len(dir) > 0 {
+			nbOfSubdir = len(strings.Split(dir, "/")) - 1
+			if e := os.MkdirAll(path.Join(m.GetPath(), dir), os.ModePerm); e != nil {
+				return e
+			}
+		}
+		if _, e := os.Readlink(path.Join(m.GetPath(), f)); e != nil {
+			// link does not exist create it
+			if e := os.Symlink(path.Join(strings.Repeat("../", nbOfSubdir)+currentDir, f), path.Join(m.GetPath(), f)); e != nil {
+				return e
+			}
+		}
+	}
+	return nil
+}
+
+// writeData writes requested metadata in specified files
+func (m *metadataVolume) writeData() error {
+	var tempDir string
+	var err error
+	tempDir, err = m.writeDataInTmpDir()
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	_, tempDirBaseName := filepath.Split(tempDir)
+	m.updateSymLinks()
+	var oldTarget string
+	oldTarget, err = os.Readlink(path.Join(m.GetPath(), currentDir))
+	if err == nil {
+		os.Remove(path.Join(m.GetPath(), currentDir)) // remove actual link
+	}
+	err = os.Symlink(tempDirBaseName, path.Join(m.GetPath(), currentDir)) // (re-)link it
+	if err != nil {
+		return err
+	}
+	if len(oldTarget) > 0 {
+		if e := os.RemoveAll(oldTarget); e != nil {
+			return e
+		}
+	}
+	return nil
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (m *metadataVolume) isDataChanged() (bool, error) {
+	errlist := []error{}
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		var data string
+		var err error
+		if data, err = fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		}
+		if isFileToGenerate(path.Join(m.GetPath(), fileName), sortStringfiedMap(data)) {
+			return true, utilErrors.NewAggregate(errlist)
+		}
+	}
+	return false, utilErrors.NewAggregate(errlist)
+}
+
+func (m *metadataVolume) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a metadata volume %v for pod %v at %v", m.volName, m.pod.UID, dir)
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := m.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, m.pod, *m.opts, m.mounter)
+	if err != nil {
+		return err
+	}
+	if err := wrapped.SetUpAt(dir); err != nil {
+		return err
+	}
+	var changed bool
+	if changed, err = m.isDataChanged(); err == nil && changed {
```

Comment:
I'm on the fence about whether we need this check.  If we just wrote out the metadata each time, the plugin would get a lot simpler.